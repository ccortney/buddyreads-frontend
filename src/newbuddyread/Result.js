import { useEffect, useState } from "react";
import {createTheme, responsiveFontSizes, ThemeProvider } from '@mui/material/styles';
import {  Typography, 
          Button, 
          Box, 
          Card, 
          CardContent, 
          CardMedia, 
          CardActions, 
          Collapse } from '@mui/material';
import BuddyReadForm from "./BuddyReadForm";
import defaultImage from "../bookcovernotfound.png"; 

/** Show limited information about a book
 * 
 * Is rendered by SearchResults to show a "card" for each book.
 * When card is clicked, new buddy read form appears. 
 *
 * Routed as /buddyreads/new
 * SearchResults -> Result -> { BuddyReadForm }
 */


const Result = ({book, formData, setFormData, handleFormChange, handleFormSubmit}) => {
  
  const [expanded, setExpanded] = useState(false);
  const [cleanedData, setCleanedData] = useState({
      id: book.id,
      title: book.volumeInfo.title, 
      authors: 'unknown', 
      publishedDate: 'unknown', 
      pageCount: null,
      image: defaultImage
  })

  let theme = createTheme();
  theme = responsiveFontSizes(theme);

  const handleClick = (book) => {
    setFormData(data => ({...data, bookId: book.id, bookTitle: book.volumeInfo.title}));
    setExpanded(!expanded);
}

  useEffect(() => {
    if (book.volumeInfo.hasOwnProperty('authors')) {
      setCleanedData(data => ({...data, authors: book.volumeInfo.authors.join(', ')}))
    }
    if (book.volumeInfo.hasOwnProperty('publishedDate')) {
      setCleanedData(data => ({...data, publishedDate: book.volumeInfo.publishedDate}))
    }
    if (book.volumeInfo.hasOwnProperty('pageCount')) {
      setCleanedData(data => ({...data, pageCount: book.volumeInfo.pageCount}))
    }
    if (book.volumeInfo.hasOwnProperty('imageLinks')) {
      setCleanedData(data => ({...data, image: book.volumeInfo.imageLinks.thumbnail}))
    }
    if (book.volumeInfo.hasOwnProperty('industryIdentifiers')) {
      setCleanedData(data => ({...data, isbn: book.volumeInfo.industryIdentifiers[0].identifier}))
    }
  }, [])


  return (

    <Card sx={{ display: 'flex' }}>
      <ThemeProvider theme={theme}>
      <CardMedia
        component="img"
        sx={{width: 90, height: 135, p: 2}}
        image={cleanedData.image}
        alt="Live from space album cover"
      />
      <Box sx={{ display: 'flex', flexDirection: 'column' }}>
      <CardContent sx={{ flex: '1 0 auto' }}>
        <Typography component="div" variant="h6" noWrap={true}>
          {cleanedData.title}
        </Typography>
        <Typography variant="subtitle1" color="text.secondary"  noWrap={true}>
          By {cleanedData.authors}
        </Typography>
        <Typography variant="subtitle1" color="text.secondary">
          {cleanedData.pageCount} pages  |  Published: {cleanedData.publishedDate}
        </Typography>
        <Typography variant="subtitle1" color="text.secondary">
          ISBN {cleanedData.isbn}
        </Typography>

      </CardContent>
      <CardActions disableSpacing>
        <Button 
          variant='contained' 
          onClick={() => handleClick(book)}
        >
          Start Buddy Read
        </Button>
      </CardActions>
      <Collapse in={expanded} timeout="auto" unmountOnExit>
        <CardContent>
          <BuddyReadForm
            formData={formData} 
            handleFormChange={handleFormChange}
            handleFormSubmit={handleFormSubmit} 
            book={book}
          />
        </CardContent>
      </Collapse>
      </Box>
      </ThemeProvider>
    </Card>
  );
}

export default Result;