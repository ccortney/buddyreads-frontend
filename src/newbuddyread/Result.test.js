import { render } from "@testing-library/react";
import Result from "./Result";

const bookInfo = {
    id: 1,
    volumeInfo: {
        title: 'title', 
        authors: ['author'], 
        publishedDate: '12/12/1912', 
        pageCount: 300
    }
}

it("renders without crashing", function() {
    render(<Result book = {bookInfo} />);
  });

it("matches snapshot", function () {
  const { asFragment } = render(<Result book = {bookInfo} />);
  expect(asFragment()).toMatchSnapshot();
});
