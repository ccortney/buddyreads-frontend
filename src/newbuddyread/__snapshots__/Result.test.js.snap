// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1oqv2c0-MuiPaper-root-MuiCard-root"
  >
    <img
      alt="Live from space album cover"
      class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-1qfy3jn-MuiCardMedia-root"
      src="bookcovernotfound.png"
    />
    <div
      class="MuiBox-root css-j7qwjs"
    >
      <div
        class="MuiCardContent-root css-1qaznj3-MuiCardContent-root"
      >
        <div
          class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-1c2nc1y-MuiTypography-root"
        >
          title
        </div>
        <h6
          class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-noWrap css-1u7magf-MuiTypography-root"
        >
          By author
        </h6>
        <h6
          class="MuiTypography-root MuiTypography-subtitle1 css-ecpxie-MuiTypography-root"
        >
          300 pages  |  Published: 12/12/1912
        </h6>
        <h6
          class="MuiTypography-root MuiTypography-subtitle1 css-ecpxie-MuiTypography-root"
        >
          ISBN 
        </h6>
      </div>
      <div
        class="MuiCardActions-root css-1rwjz6-MuiCardActions-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Start Buddy Read
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
